sdk/schemaregistry/Azure.Data.SchemaRegistry/src/Serialization/SchemaRegistrySerializer.cs

```diff
@@ -0,0 +1,554 @@
+        /// <param name="cancellationToken">An optional <see cref="CancellationToken"/> instance to signal the request to cancel the operation.</param>
+        /// <exception cref="InvalidOperationException">
+        ///   This can occur if the <c>groupName</c> was not specified when constructing the <see cref="SchemaRegistrySerializer"/>.
+        ///   It can also occur if the <paramref name="messageType"/> does not have a public parameterless constructor.
+        /// </exception>
+        /// <exception cref="RequestFailedException">
+        ///   An error occurred while attempting to communicate with the Schema Registry service. This can occur if the schema generated from the type
+        ///   does not exist in the registry.
+        /// </exception>
+        /// <exception cref="Exception">
+        ///   The data did not adhere to the specified schema, or the schema itself was invalid.
+        ///   The <see cref="Exception.InnerException"/> will contain the underlying exception from the <see cref="SchemaValidator{T}"/> class.
+        /// </exception>
+        public async ValueTask<MessageContent> SerializeAsync(
+            object data,
+            Type dataType = default,
+            Type messageType = default,
+            CancellationToken cancellationToken = default)
+            => await SerializeInternalAsync(data, dataType, messageType, true, cancellationToken).ConfigureAwait(false);
+
+        internal async ValueTask<MessageContent> SerializeInternalAsync(
+            object data,
+            Type dataType,
+            Type messageType,
+            bool async,
+            CancellationToken cancellationToken)
+        {
+            if (_groupName == null)
+            {
+                throw new InvalidOperationException(
+                    "A group name must be specified in the 'SchemaRegistryJsonSerializer' constructor if you will be attempting to serialize. " +
+                    "The group name can be omitted if only deserializing.");
+            }
+
+            messageType ??= typeof(MessageContent);
+            if (messageType.GetConstructor(Type.EmptyTypes) == null)
+            {
+                throw new InvalidOperationException(
+                    $"The type {messageType} must have a public parameterless constructor in order to use it as the 'MessageContent' type to serialize to.");
+            }
+
+            var message = (MessageContent)Activator.CreateInstance(messageType);
+
+            (string retrievedSchemaId, BinaryData bd) = async
+                ? await SerializeInternalAsync(data, dataType, true, cancellationToken).ConfigureAwait(false)
+                : SerializeInternalAsync(data, dataType, false, cancellationToken).EnsureCompleted();
+
+            message.Data = bd;
+            var mimeType = (_serializerOptions.Format == SchemaFormat.Json) ? JsonMimeType : CustomMimeType;
```

Comment:
Oh I missed that we are only allowing Json or custom. Is there a reason for this given we allow specifying any format?

Reply:
I guess my thinking was that Json/custom covers everything but Avro, which I'm assuming most people wouldn't choose this serializer to use with Avro. I can update this to pull the first piece from the ContentType https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/schemaregistry/Azure.Data.SchemaRegistry/src/Generated/Models/ContentType.cs#L14

Reply:
SchemaFormat is extensible so technically users can pass in any value they want.

As for Avro, the content type for an Avro payload is defined here - https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/schemaregistry/Microsoft.Azure.Data.SchemaRegistry.ApacheAvro/src/SchemaRegistryAvroSerializer.cs#L29